yellow-paper/docs/gas-and-fees/gas-metering-and-fees.md

```diff
@@ -0,0 +1,409 @@
+
+This amount must be initialized for each public kernel iteration as:
+
+```
+daGasLeft := public_kernel_public_inputs.da_gas_limit 
+             - public_kernel_public_inputs.end_non_revertible.da_gas_used
+             - public_kernel_public_inputs.end.da_gas_used
+```
+
+At the end of execution, the public VM will have an updated `daGasLeft` value.
+
+This value will be used by the public kernel to update the `da_gas_used` in the `PublicAccumulatedData` struct corresponding to the current phase:
+
+```
+# if in setup 
+public_kernel_public_inputs.end_non_revertible.gas_used.da_gas_used := 
+        public_kernel_public_inputs.gas_limits.da_gas_limit
+          - daGasLeft
+          - public_kernel_public_inputs.end.gas_used.da_gas_used
+
+# if in app logic
+public_kernel_public_inputs.end.gas_used.da_gas_used := 
+        public_kernel_public_inputs.gas_limits.da_gas_limit
+          - daGasLeft
+          - public_kernel_public_inputs.end_non_revertible.gas_used.da_gas_used
+```
+
+
+It will be the same for the `compute_gas_used`.
+
+### Out of gas
+
+Public execution must stop as soon as possible once `daGasLeft` is exhausted.
+
+The public VM will revert if an exception is thrown, including an out of gas exception.
+When this occurs, the public VM will return a revert code.
+
+If this occurs in setup, the transaction is invalid and cannot be included in the block.
+
+If this occurs in app_logic, `daGasLeft` is set to 0, and thus `public_kernel_public_inputs.gas_limits.da_gas_limit == public_kernel_public_inputs.end.gas_used.da_gas_used + public_kernel_public_inputs.end_non_revertible.gas_used.da_gas_used`. 
+
+Similarly for `computeGasLeft`.
+
+In the teardown phase, there cannot be an out of gas error because the teardown phase effectively has infinite gas (since the user has prepaid for it via `inclusion_fee`). If there is any other exception though, the transaction is invalid and cannot be included in the block. It is thus expected that sequencers will maintain a whitelist of teardown functions and the minimum `inclusion_fee` required for each.
+
+### Recombining for teardown
+
+The public kernel **tail** will recombine the `gas_used` values from the non-revertible and revertible phases.
+
+```
+kernel_public_inputs.gas_limits := public_kernel_public_inputs.gas_limits
+
+kernel_public_inputs.end.gas_used.da_gas_used =
+    public_kernel_public_inputs.end_non_revertible.gas_used.da_gas_used
+    + public_kernel_public_inputs.end.gas_used.da_gas_used
+
+kernel_public_inputs.end.gas_used.compute_gas_used :=
+    public_kernel_public_inputs.end_non_revertible.gas_used.compute_gas_used +
+    + public_kernel_public_inputs.end.gas_used.compute_gas_used
+
+
+fee_per_da_gas := public_kernel_public_inputs.constants.historical_header.global_variables.fee_per_da_gas
+fee_per_compute_gas := public_kernel_public_inputs.constants.historical_header.global_variables.fee_per_compute_gas
```

Comment:
Isn't `historical_header` the value from the _previous_ block? Do we want to use those values, or new ones set for the current block?